File: Algorithms/python/019 - Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        arr = []
        it = head
        
        while it:
            arr.append(it)
            it = it.next
        
        index = len(arr) - n

        if index == 0:
            head = head.next
        else:
            arr[index - 1].next = arr[index].next
        return head

File: cpp/019 - Remove Nth Node From End of List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // var
        ListNode* dummy = new ListNode(0);
        ListNode* faster = dummy;
        ListNode* slower = dummy;
        
        // init
        dummy->next = head;
        for (int i = 0; i <= n; i++)
            faster = faster->next;
        
        while (faster) {
            faster = faster->next;
            slower = slower->next;
        }
        
        // remove nth node
        slower->next = slower->next ? slower->next->next : NULL;
        
        return dummy->next;
    }
};